### PR TITLE
User list

### LINKED ISSUE #68: User list UI
Would be nice to have a list of users and their online/offline states on the right side of the screen a la every other chat application in existence!

###### Once again @PullJosh if you write some HTML/CSS for this I'll make it actually function.

### PR DESCRIPTION
**Pull request by towerofnix**
Demo screenshot

Fixes #68.

User list is sorted by prioritizing online users, then alphabetically (A at top).

See also [PullJosh's UI reference](https://codepen.io/PullJosh/pen/OOYZWK). Most new CSS is copied from there, and I reused a lot of the sidebar's CSS (just by including it as another class prefix, not by copy/pasting it!).

@heyitsmeuralex Not sure what semver this is?

### CODE REVIEW

**Comment by heyitsmeuralex** [comment]
If we're going to make offline users opaque, the online/offline indicator circle doesn't really have a purpose anymore. Might I suggest that gets removed?

**Comment by heyitsmeuralex** [comment]
Semver for this would be minor, because it **adds functionality in a backwards-compatible manner**.

File: packages/client/src/app.js [hunk 1]
```diff
@@ -54,13 +55,16 @@ app.use((state, emitter) => {
 
   state._session = new Proxy({}, {
     set: function(target, key, value) {
-      if (key === 'id') {
+      const ret = Reflect.set(target, key, value)
```
Comment (heyitsmeuralex) [inline packages/client/src/app.js]: Woah, what does this do? ...returns a boolean if `target[key] = value` was successful...? wat?

Conclusion: Proxy and Reflect are both magic. I realise this existed before but never noticed it's usage. thisandagain pls explain.

File: packages/client/src/app.js [hunk 2]
```diff
@@ -152,6 +156,7 @@ app.use((state, emitter) => {
 
 app.use(messages.store)
 app.use(sidebar.store)
+app.use(userList.store)
 app.use(accountSettings.store)
```
Comment (heyitsmeuralex) [inline packages/client/src/app.js]: aaaa I should really get started on fixing data flow in the app because this is bad

**Comment by towerofnix** [comment]
Sounds good, removed the bubble.

File: packages/client/src/components/user-list.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+              return +1
+            }
+
+            if (a.username < b.username) {
+              return -1
+            }
+
+            if (b.username < a.username) {
+              return +1
+            }
```
Comment (heyitsmeuralex) [inline packages/client/src/components/user-list.js]: Could be worth formatting this as

```js
.sort((a, b) => do {
  if (cond) -1
  if (cond) +1
  0
})
```

Wait, no, `do {}` isn't in Stage 3 yet. Nevermind.
If we ever end up doing babel - probably when I replace choo with preact - using its do expression transform would be awesome (especially for (JSX/bel)related things!)

File: packages/client/src/components/user-list.js [hunk 4]
```diff
@@ -97,7 +97,10 @@ const component = (state, emit) => {
             return 0
           }).map(user => {
             return html`
-              <div class='item user ${user.online ? 'online' : 'offline'}'>
+              <div
+                class='item user ${user.online ? 'online' : 'offline'}'
+                title='${user.username} (${user.online ? 'Online' : 'Offline'})'
```
Comment (heyitsmeuralex) [inline packages/client/src/components/user-list.js]: :heart:

**Comment by towerofnix** [comment]
I'm actually wondering if they maybe should be kept, though -- profile pictures which inherently aren't that saturated can look like they're offline, and then the only way to tell whether someone's online or not is by checking how dark their username text is.

**Comment by towerofnix** [comment]
Comparison screenshot:

**Comment by heyitsmeuralex** [comment]
Honestly I think the opacity of the username itself gives it away quite clearly, especially due to how the list is sorted.

**Comment by towerofnix** [comment]
Alright! If it's a huge problem in the future we can always change it back.

**Comment by heyitsmeuralex** [comment]
@towerofnix just added issue label descriptions for everything including the `semver` ones, so you should have a slightly less difficult time figuring out if it's major, minor, or a patch.

**Comment by towerofnix** [comment]
@heyitsmeuralex Huh, cool! I didn't know label descriptions were a thing. Definitely handy.

**Comment by heyitsmeuralex** [comment]
Pretty sure they're a very new thing.

**Pull request merged**